cardano-cli/src/Cardano/CLI/Run/Debug/CheckNodeConfiguration.hs

```diff
@@ -0,0 +1,188 @@
+      pure $
+        Aeson.insert (Aeson.fromText $ eraToGenesisHashKey era) (Aeson.String actualHash) nodeConfigObj
+ where
+  configFilePath = unFile configFile
+
+-- | Build an 'Analysis' value from the given node configuration file, for the given era.
+-- Reads the genesis file path and hash from the node configuration file, if they exist.
+mkAnalysis
+  :: AnyCardanoEra
+  -- ^ The era to analyze
+  -> NodeConfigFile 'In
+  -- ^ The node configuration file path. It's not read by this function, but used for producing error messages.
+  -> NodeConfig
+  -- ^ The parsed node configuration file
+  -> ExceptT DebugCmdError IO (Maybe Analysis)
+mkAnalysis cEra@(AnyCardanoEra era) configFile nodeConfig = do
+  case (getData era FileRole nodeConfig, getData era HashRole nodeConfig) of
+    (Nothing, Nothing) ->
+      -- Neither the genesis file nor the genesis hash are declared in the node configuration file
+      pure Nothing
+    (Nothing, Just _) -> do
+      -- The path to the genesis file is not given in the node configuration file,
+      -- but a genesis hash is provided. Since we can't verify the latter,
+      -- we write something to stdout, because this is weird.
+      liftIO $
+        putStrLn $
+          "Genesis file for " <> Text.unpack (eraToStringKey era) <> " is not declared in " <> configFilePath
+      liftIO $ putStrLn "So the hash's correctness cannot be checked."
+      pure Nothing
+    (Just filepath, expectedHash) -> do
+      -- We make the genesis filepath relative to the node configuration file, like the node does:
+      -- https://github.com/IntersectMBO/cardano-node/blob/9671e7b6a1b91f5a530722937949b86deafaad43/cardano-node/src/Cardano/Node/Configuration/POM.hs#L668
+      -- Note that, if the genesis filepath is absolute, the node configuration file's directory is ignored (by property of </>)
+      let genesisFilePath = takeDirectory configFilePath </> filepath
+      actualHash <-
+        case era of
+          ByronEra -> do
+            (_, GenesisHash byronHash) <-
+              firstExceptT (DebugNodeConfigGenesisDataCmdError genesisFilePath) $ readGenesisData genesisFilePath
+            return $ Text.pack $ show byronHash
+          _ -> Crypto.hashToTextAsHex <$> Read.readShelleyOnwardsGenesisAndHash genesisFilePath
+      case expectedHash of
+        Nothing ->
+          pure $ Just $ Analysis cEra genesisFilePath actualHash Nothing
+        Just expectedHash' ->
+          pure $ Just $ Analysis cEra genesisFilePath actualHash (Just expectedHash')
+ where
+  configFilePath = unFile configFile
+
+data FileRole
+
+data HashRole
+
+-- | A GADT, to be able to cover both files and hashes in a single function, e.g. 'getData' below
+data Role r where
+  FileRole :: Role FileRole
+  HashRole :: Role HashRole
+
+-- | A type family, to be able to cover both files and hashes in a single function, e.g. 'getData' below
+type family DataKind r where
```

Comment:
This is overkill.

Reply:
Removed :+1: (it was the simplest use of a `type family` though :wink:)